### PR TITLE
Blackbox

### PR DESCRIPTION
**Pull request by joelin0**
Added a framework for producing general blackbox programs. Implemented Deutsch-Jozsa, Bernstein-Vazirani and Simon's algorithms using this framework.

### CODE REVIEW

File: grove/blackbox_algorithms/bernstein_vazirani.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+"""Module for the Bernstein-Vazirani Algorithm."""
```
Comment (ampolloreno) [inline grove/blackbox_algorithms/bernstein_vazirani.py]: Can you add a reference for notes that helped you write this?

File: grove/blackbox_algorithms/bernstein_vazirani.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+class BernsteinVaziraniAlgorithm(AbstractBlackBoxAlgorithm):
+    def __init__(self, n, vec_a, b):
+        """
+        Creates an instance of an object to simulate the Bernstein-Vazirani Algorithm
+        such that the function f(x) is given by a * x + b, where a is a bitwise dot product and all additions are mod 2.
+        :param n: the number of bits in the domain of the function
+        :param vec_a: a length n vector of 0s and 1s
+        :param b: a binary bit, either 0 or 1
+        """
```
Comment (ampolloreno) [inline grove/blackbox_algorithms/bernstein_vazirani.py]: missing `:return:` and `:rtype:`, among other things.

Please reference, for example, [here](https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html) for reST docstrings, and please apply to the rest of the docstrings.

File: grove/blackbox_algorithms/deutsch_jozsa.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
+"""Module for Deutsch-Jozsa Algorithm. Adapted from Harley Patton's deutsch_jozsa.py"""
+import pyquil.quil as pq
+from pyquil.gates import *
+import utils as bbu
+from blackbox import AbstractBlackBoxAlgorithm
+
+class DeutschJozsa(AbstractBlackBoxAlgorithm):
```
Comment (ampolloreno) [inline grove/blackbox_algorithms/deutsch_jozsa.py]: If we are going to merge in @harleypatton  `DeutschJozsa`, we shouldn't code review this twice. If this is the way you all think it should be written, this should be broken off into its own PR dependent on `AbstractBlackBoxAlgorithm`.

File: grove/blackbox_algorithms/tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
+"""Test class for methods found in the black box utils."""
+
+from grove.blackbox_algorithms.utils import *
+import pyquil.quil as pq
+from pyquil.resource_manager import Qubit
+import unittest
+
+
+class TestUnitaryFromFunction(unittest.TestCase):
```
Comment (ampolloreno) [inline grove/blackbox_algorithms/tests/test_utils.py]: The rest of Grove/Pyquil uses `pytest` rather than `unittest`. It would be much better if we could keep all of our codebase on the same testing package. Can you refactor these to use `py.test` instead?

File: grove/blackbox_algorithms/deutsch_jozsa.py (hunk 3 above)
Comment (ncrubin) [inline grove/blackbox_algorithms/deutsch_jozsa.py]: @harleypatton thoughts?

Comment (harleypatton) [inline grove/blackbox_algorithms/deutsch_jozsa.py]: @ncrubin I just talked with @joelin0 and we think we will keep these algorithms separate for now. I believe he is replacing this pull request soon. Let us know if you have any input on this.

**Comment by joelin0** [comment]
@ampolloreno @pkaralekas @ncrubin @harleypatton Harley and I decided it would be best to keep these algorithms all separated. I have made PR #39 for Bernstein-Vazirani and PR #40 for Simon's algorithm. Harley's Deutsch-Jozsa remains at PR #26.
